p2p/transport/webrtc/listener.go

```diff
@@ -0,0 +1,310 @@
+}
+
+func (l *listener) Multiaddrs() []ma.Multiaddr {
+	return []ma.Multiaddr{l.localMultiaddr}
+}
+
+func (l *listener) accept(ctx context.Context, addr candidateAddr) (tpt.CapableConn, error) {
+	var (
+		scope network.ConnManagementScope
+		pc    *webrtc.PeerConnection
+	)
+
+	cleanup := func() {
+		if scope != nil {
+			scope.Done()
+		}
+		if pc != nil {
+			_ = pc.Close()
+		}
+	}
+
+	remoteMultiaddr, err := manet.FromNetAddr(addr.raddr)
+	if err != nil {
+		return nil, err
+	}
+
+	scope, err = l.transport.rcmgr.OpenConnection(network.DirInbound, false, remoteMultiaddr)
+	if err != nil {
+		defer cleanup()
+		return nil, err
+	}
+
+	settingEngine := webrtc.SettingEngine{}
+	settingEngine.SetAnsweringDTLSRole(webrtc.DTLSRoleServer)
+	settingEngine.SetICECredentials(addr.ufrag, addr.ufrag)
+	settingEngine.SetLite(true)
+	settingEngine.SetICEUDPMux(l.mux)
+	settingEngine.DisableCertificateFingerprintVerification(true)
+	settingEngine.DetachDataChannels()
+
+	api := webrtc.NewAPI(webrtc.WithSettingEngine(settingEngine))
+
+	pc, err = api.NewPeerConnection(l.config)
+	if err != nil {
+		defer cleanup()
+		return nil, err
+	}
+
+	// signaling channel wraps an error in a struct to make
+	// the error nullable.
+	signalChan := make(chan struct{ error })
+	var wrappedChannel *dataChannel
+	var handshakeOnce sync.Once
+	// this enforces that the correct data channel label is used
+	// for the handshake
+	handshakeChannel, err := pc.CreateDataChannel("handshake", &webrtc.DataChannelInit{
+		Negotiated: func(v bool) *bool { return &v }(true),
+		ID:         func(v uint16) *uint16 { return &v }(0),
+	})
+	if err != nil {
+		defer cleanup()
+		return nil, err
+	}
+
+	// The raw datachannel is wrapped in the libp2p abstraction
+	// as early as possible to allow any messages sent by the remote
+	// to be buffered. This is done since the dialer leads the listener
+	// in the handshake process, and a faster dialer could have set up
+	// their connection and started sending Noise handshake messages before
+	// the listener has set up the onmessage callback. In this use case,
+	// since the data channels are negotiated out-of-band, they will be
+	// instantly in `readyState=open` once the SCTP connection is set up.
+	// Therefore, we wrap the datachannel before performing the
+	// offer-answer exchange, so any messages sent from the remote get
+	// buffered.
+	handshakeChannel.OnOpen(func() {
+		log.Debugf("handshake channel open")
+		rwc, err := handshakeChannel.Detach()
+		if err != nil {
+			signalChan <- struct{ error }{errDatachannel("could not detach", err)}
+			return
+		}
+		wrappedChannel = newDataChannel(
+			handshakeChannel,
+			rwc,
+			pc,
+			l.mux.LocalAddr(),
+			addr.raddr,
+		)
+		handshakeOnce.Do(func() {
+			signalChan <- struct{ error }{nil}
+		})
+	})
+
+	// Checking the peerconnection state is not necessary in this case as any
+	// error caused while accepting will trigger the onerror callback of the
+	// handshake channel.
```

Comment:
There is an additional check possible where we wait for the `peerConnection` to move into the connected state and we can use a callback to await this condition. Here, if the peerconnection fails to connect, it will trigger the onerror callback of the datachannel. Therefore the check would be unnecessary.